src/main/java/com/cst438/service/RegistrarServiceProxy.java

```diff
@@ -0,0 +1,175 @@
+                    sectionRepository.save(s);
+                }
+
+            } else if (key.equals("deleteSection")) {
+                sectionRepository.deleteById(Integer.parseInt(parts[1]));
+
+            } else if (key.equals("updateSection")) {
+                SectionDTO dto = fromJsonString(parts[1], SectionDTO.class);
+                Section s = sectionRepository.findById(dto.secNo()).orElse(null);
+                if (s != null) {
+                    Course c = courseRepository.findById(dto.courseId()).orElse(null);
+                    if (c != null) {
+                        s.setCourse(c);
+                        s.setSecId(dto.secId());
+                        s.setBuilding(dto.building());
+                        s.setRoom(dto.room());
+                        s.setTimes(dto.times());
+                        s.setInstructor_email(dto.instructorEmail());
+                        sectionRepository.save(s);
+                    }
+                }
+
+            } else if (key.equals("addUser")) {
+                UserDTO dto = fromJsonString(parts[1], UserDTO.class);
+                User u = new User();
+                u.setId(dto.id());
+                u.setName(dto.name());
+                u.setEmail(dto.email());
+                u.setType(dto.type());
+                userRepository.save(u);
+
+            } else if (key.equals("deleteUser")) {
+                userRepository.deleteById(Integer.parseInt(parts[1]));
+
+            } else if (key.equals("updateUser")) {
+                UserDTO dto = fromJsonString(parts[1], UserDTO.class);
+                User u = userRepository.findById(dto.id()).orElse(null);
+                if (u != null) {
+                    u.setName(dto.name());
+                    u.setEmail(dto.email());
+                    u.setType(dto.type());
+                    userRepository.save(u);
+                }
+
+            } else if (key.equals("addEnrollment")) {
+                EnrollmentDTO dto = fromJsonString(parts[1], EnrollmentDTO.class);
+                Enrollment e = new Enrollment();
+                e.setEnrollmentId(dto.enrollmentId());
+                Section s = sectionRepository.findById(dto.sectionNo()).orElse(null);
+                User u = userRepository.findById(dto.studentId()).orElse(null);
+                if (s != null && u != null) {
+                    e.setSection(s);
+                    e.setStudent(u);
+                    e.setGrade(dto.grade());
+                    enrollmentRepository.save(e);
+                }
+
+            } else if (key.equals("deleteEnrollment")) {
+                enrollmentRepository.deleteById(Integer.parseInt(parts[1]));
+
+            } else {
+                System.out.println("Invalid action: " + key);
+            }
+
+        } catch (Exception e) {
+            System.out.println("Exception in receivedFromRegistrar " + e.getMessage());
+        }
+    }
```

Comment:
Correctly listens for the various action methods in the registrar's code. They correspond 1 send and 1 receive.